Fallback if Feedback submission fails. Help user classify NAT if proxying fails.

lucyhe opened the pull request:
To do in this pull request:
- Add 'X' button to close Unable to Proxy dialog (or add 'Try Again' option beside the 'Analyze Network' button)
- Add padding on top of 'Submit Feedback' on the Splash page
- Make the checkbox label on the feedback page link to the View Logs page, and remove the pop-out icon that currently opens that page
- Add a timeout to getNatType when feedback is being sent -- if it takes longer than 30s (?) we will submit all feedback except the NAT type. This will guarantee that the 'Submitting Feedback' dialog will be closed.

To do in future pull request: the new uproxy-troubleshoot element should be added to the Advanced Settings page when that is created (the contents of uproxy-troubleshoot should be taken out of the paper-dialog and added to a core-overlay)

![image](https://cloud.githubusercontent.com/assets/8723127/6914091/0bc64b34-d755-11e4-8c5b-13c96e5048b6.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6914122/5a720016-d755-11e4-879f-762838c2e2bd.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6913330/b947d182-d74c-11e4-81b0-06765abb10b8.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6913589/45c06644-d750-11e4-87a4-7defb4b34c05.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6913592/49e9b518-d750-11e4-9e23-c32267eb535d.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6913596/51b76434-d750-11e4-8d82-5f43699ac7ad.png)


<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1180)
<!-- Reviewable:end -->

lucyhe commented:
Adding X to close dialog:
![image](https://cloud.githubusercontent.com/assets/8723127/6926045/a39eed7e-d7b1-11e4-8e4f-ca829faba2b0.png)

salomegeo commented:
What happens if it's not symmertric?

On Tue, Mar 31, 2015 at 2:24 PM, lucyhe <notifications@github.com> wrote:

> Reopened #1180 <https://github.com/uProxy/uproxy/pull/1180>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/uProxy/uproxy/pull/1180#event-269865214>.
>

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)

Reviewed files:
- src/generic_core/core.ts @ r1
- src/generic_ui/polymer/feedback.html @ r1
- src/generic_ui/polymer/feedback.ts @ r1
- src/generic_ui/polymer/instance.ts @ r1
- src/generic_ui/polymer/root.html @ r1
- src/generic_ui/polymer/troubleshoot.html @ r1
- src/generic_ui/polymer/troubleshoot.ts @ r1
- src/generic_ui/scripts/core_connector.ts @ r1
- src/uproxy.ts @ r1





---

<sup>**[src/generic_core/core.ts, line 438 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllF2Q6jgizhkuEa1Cy)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L438)):</sup>
Personal preference (feel free to ignore this):

Since we are not using any of these variables anywhere else, we could also just rewrite the below function as:

```
return new Promise((fulfill, reject) => {
...
   fulfill();
...
  reject();
...
});
```

Alternatively:

```
var fulfill :Function;
var reject :Function;

... {the body you have already} ...

return new Promise((f, r) => {
  fulfill = f;
  reject = r;
});
```

just so we don't have more properties on the object we are not using.

---

<sup>**[src/generic_core/core.ts, line 450 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllIBh_Gq3EvXaHWPIc)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L450)):</sup>
Is there an enum you could reference for the state and status values?

---

<sup>**[src/generic_core/core.ts, line 453 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllG83e6OcEfhHU__k3)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L453)):</sup>
Can you make an issue for this and add it to the comment?

---

<sup>**[src/generic_core/core.ts, line 472 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllGYCltcrvolpAa102)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L472)):</sup>
Possible cleaner way to do this (pseudocode):

```
var extradata = [];

if feedback.networkInfo
  extradata.push this.getNetworkInfo()

if feedback.logs
  extradata.push this.getLogs()

return Promise.all(extradata).then (data) ->
  return sendXhr(data.join('\n')) # this could be inlined if you wanted
```

Not sure if this would actually end up being cleaner, but I think it might get the point across a little better and make it easier to add more data in the future

---

<sup>**[src/generic_ui/polymer/feedback.html, line 147 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllIZW5O6BMjYW2X2vk)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_ui/polymer/feedback.html#L147)):</sup>
What is this div for?

---

<sup>**[src/generic_ui/polymer/instance.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllIsmTUdF4ye1Fw4bI)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_ui/polymer/instance.ts#L28)):</sup>
Completely and totally personal preference (again, feel free to ignore): I think it looks a bit better if arguments to a function that need to be on a second line have spacing a bit farther than the opening paren.

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
@salomegeo 
The possibilities are:
- NAT type detected: **SymmetricNAT** This NAT type is **very likely** restricting your ability to use uProxy.
- NAT type detected: **PortRestrictedCone** This NAT type is **possibly** restricting your ability to use uProxy.
- NAT type detected: **RestrictedCone** This NAT type is **unlikely** restricting your ability to use uProxy.
- NAT type detected: **FullCone** This NAT type is **unlikely** restricting your ability to use uProxy.

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)

Reviewed files:
- src/generic_core/diagnose.ts @ r2
- src/generic_ui/polymer/feedback.html @ r2
- src/generic_ui/polymer/splash.html @ r2
- src/generic_ui/polymer/troubleshoot.html @ r2
- src/generic_ui/scripts/core_connector.ts @ r2





---

<sup>**[src/generic_core/core.ts, line 505 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllNnx1vK4HoP8OBBzj)** ([raw file](https://github.com/uproxy/uproxy/blob/dc8567c5aa7b5c95b3b733ef064e43fd181b5150/src/generic_core/core.ts#L505)):</sup>
It almost seems like aborting after a timeout should be in `doNatProvoking`, totally fine with this if that is not feasible though.

---

<sup>**[src/generic_ui/polymer/feedback.html, line 169 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllOjHjUL4hb-WMMeh1)** ([raw file](https://github.com/uproxy/uproxy/blob/dc8567c5aa7b5c95b3b733ef064e43fd181b5150/src/generic_ui/polymer/feedback.html#L169)):</sup>
Unintended comment?

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)

Reviewed files:
- src/generic_core/core.ts @ r2





---



<!-- Sent from Reviewable.io -->

salomegeo commented:
But the rest is the same right?

On Tue, Mar 31, 2015 at 3:57 PM, lucyhe <notifications@github.com> wrote:

> @salomegeo <https://github.com/salomegeo>
> The possibilities are:
>
>    - NAT type detected: *SymmetricNAT* This NAT type is *very likely*
>    restricting your ability to use uProxy.
>    - NAT type detected: *PortRestrictedCone* This NAT type is *possibly*
>    restricting your ability to use uProxy.
>    - NAT type detected: *RestrictedCone* This NAT type is *unlikely*
>    restricting your ability to use uProxy.
>    - NAT type detected: *FullCone* This NAT type is *unlikely*
>    restricting your ability to use uProxy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/uProxy/uproxy/pull/1180#issuecomment-88227436>.
>

lucyhe commented:
@salomegeo for the "Unable to proxy" dialog everything else is the same!

lucyhe commented:
More UI changes after chatting with @zahoriea:
- Instead of an X, give users a "Try Reconnecting" option.
- Clicking 'Analyze Network' opens up the view-logs page
- Opening up uproxy again shows that the network has been analyzed.
If users click 'Submit Feedback' without first analyzing their network, the feedback page is opened **without** "include logs" checked. If they've already analyzed their network, "include logs" will be checked.

![image](https://cloud.githubusercontent.com/assets/8723127/6932328/c0e3222c-d7e5-11e4-8b38-20e2839d7c4a.png)
![image](https://cloud.githubusercontent.com/assets/8723127/6932364/4d3eec88-d7e6-11e4-8405-354138d27268.png)
![image](https://cloud.githubusercontent.com/assets/8723127/6932391/a17e8bfa-d7e6-11e4-882d-feb039e7d536.png)
![image](https://cloud.githubusercontent.com/assets/8723127/6932394/a7a93066-d7e6-11e4-993b-1ded433c2de8.png)

jpevarnek commented:
@lucyhe my instinct is that "Try Reconnecting" would immediately try to start a connection with that person again.  Is there a way to exit from the failure screen and not try reconnecting (if not, I think there should be)?

lucyhe commented:
@jpevarnek ah, you're right, it does seem a little misleading now. switched try connecting to an 'X' at the top again, and might add try reconnecting later

![image](https://cloud.githubusercontent.com/assets/8723127/6932900/10b14920-d7ee-11e4-8244-cd8139baefe6.png)

![image](https://cloud.githubusercontent.com/assets/8723127/6932890/dbb500cc-d7ed-11e4-8704-2722317109d9.png)

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)







---

<sup>**[src/generic_core/core.ts, line 472 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllGYCltcrvolpAa102)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L472)):</sup>
This sounds good. For now I ended up taking out the networkInfo boolean (not sure that we will have users submitting just nat type), but I will keep this in mind if it's added later

---

<sup>**[src/generic_core/core.ts, line 505 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllNnx1vK4HoP8OBBzj)** ([raw file](https://github.com/uproxy/uproxy/blob/dc8567c5aa7b5c95b3b733ef064e43fd181b5150/src/generic_core/core.ts#L505)):</sup>
I think since the timeout is introduced for UX purposes (so the user doesn't feel "stuck" if it takes a long time), it might be best to leave the original function as is (since diagnose.ts doesn't really have time constraints)

---

<sup>**[src/generic_ui/polymer/feedback.html, line 146 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllIZW5O6BMjYW2X2vk)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_ui/polymer/feedback.html#L146)):</sup>
Not needed, removed!

---

<sup>**[src/generic_ui/polymer/feedback.html, line 169 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllOjHjUL4hb-WMMeh1)** ([raw file](https://github.com/uproxy/uproxy/blob/dc8567c5aa7b5c95b3b733ef064e43fd181b5150/src/generic_ui/polymer/feedback.html#L169)):</sup>
Removed

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)







---

<sup>**[src/generic_core/core.ts, line 438 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllF2Q6jgizhkuEa1Cy)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L438)):</sup>
Done.

---

<sup>**[src/generic_core/core.ts, line 450 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllIBh_Gq3EvXaHWPIc)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L450)):</sup>
Changed 4 to XMLHttpRequest.DONE and added a comment for 200.

---

<sup>**[src/generic_core/core.ts, line 453 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JllG83e6OcEfhHU__k3)** ([raw file](https://github.com/uproxy/uproxy/blob/9c005aaad091096808d5d434410a15ca985ee374/src/generic_core/core.ts#L453)):</sup>
Done.

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)

Reviewed files:
- src/generic_core/core.ts @ r5
- src/generic_ui/polymer/feedback.html @ r5
- src/generic_ui/polymer/feedback.ts @ r4
- src/generic_ui/polymer/instance.ts @ r3
- src/generic_ui/polymer/logs.html @ r6
- src/generic_ui/polymer/logs.ts @ r3
- src/generic_ui/polymer/troubleshoot.html @ r4
- src/generic_ui/polymer/troubleshoot.ts @ r4
- src/uproxy.ts @ r4





---

<sup>**[src/generic_core/core.ts, line 439 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JlqD1G3WDGhnhMCSFmZ)** ([raw file](https://github.com/uproxy/uproxy/blob/a8b06512b7918e6052495c265cc52b9454c0cb11/src/generic_core/core.ts#L439)):</sup>
I think naming these `fulfill` and `reject` would be a bit more clear to anyone casually looking at the file.

---

<sup>**[src/generic_core/core.ts, line 440 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JlqIDm6QFhiAZ4eoWeb)** ([raw file](https://github.com/uproxy/uproxy/blob/a8b06512b7918e6052495c265cc52b9454c0cb11/src/generic_core/core.ts#L440)):</sup>
I think it might be cleaner to have this as a private method instead of inlining it.

---

<sup>**[src/generic_core/core.ts, line 501 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JlqHGDSEdiSKu4Yj_EK)** ([raw file](https://github.com/uproxy/uproxy/blob/a8b06512b7918e6052495c265cc52b9454c0cb11/src/generic_core/core.ts#L501)):</sup>
Thinking about this a bit more, I think there is one weird case here:

0. Request 1 arrives
0. 10 seconds pass
0. Request 2 arrives
0. x seconds pass
0. Request 2 completes
0. 20 - x seconds pass
0. Request 1 times out

In this case, for the first time we tried to get the NAT type, we will return that the classification timed out even if we already have a valid cached value.

I'll completely defer to you on what's expected in this case, or if this is even a problem at all.

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)







---

<sup>**[src/generic_core/core.ts, line 439 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JlqD1G3WDGhnhMCSFmZ)** ([raw file](https://github.com/uproxy/uproxy/blob/a8b06512b7918e6052495c265cc52b9454c0cb11/src/generic_core/core.ts#L439)):</sup>
Done.

---

<sup>**[src/generic_core/core.ts, line 501 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/1180#-JlqHGDSEdiSKu4Yj_EK)** ([raw file](https://github.com/uproxy/uproxy/blob/a8b06512b7918e6052495c265cc52b9454c0cb11/src/generic_core/core.ts#L501)):</sup>
Oh interesting! I don't think it's expected for an earlier call to return the output of a later call, so I don't feel too strongly about checking for this case. 

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
LGTM

---

Reviewed files:
- src/generic_core/core.ts @ r7

---



---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1180)
<!-- Sent from Reviewable.io -->
